# 感知机（Perceptron）神经网络实现

Renato Barros

## 引言

在现代商业环境中，做出敏锐而精准的决策是企业取得成功的关键。在这一背景下，
感知机（Perceptron）作为一种受生物神经元启发的机器学习算法，能够为数据分析
与行为预测提供强有力的工具支持。

感知机就像一个专注于学习的学徒，通过观察数据中的规律与关联，逐步学会做出
预测并辅助业务决策。无论是识别高潜力热销商品、防范欺诈风险，还是实现个性化
客户体验，感知机都能在推动企业智能化与效率提升方面发挥重要作用。

## 工作原理

感知机的本质是一个基础的人工神经网络：接收输入特征、进行加权处理并输出结果。
该过程主要包含以下简单的数学步骤：

1.  **线性组合（加权求和）：** 每个输入特征（数值）均会乘以对应的权重，权重代表了
    该特征在最终决策中的重要程度。所有乘积相加后，得到一个中间值，称为“加权和”。

    公式：`u = (x1 * 权重1) + (x2 * 权重2) + ...`

2.  **激活函数：** 上一步计算出的加权和 (u) 会传递给激活函数，由激活函数决定
    感知机的最终输出。激活函数为模型引入了阈值判定机制。在本例中，我们采用“阶跃函数”
    （Step Function），其功能类似于逻辑开关：

    公式：

    `f(u) = 1 (若 u >= 0) 否则 0`

    其中：

    -   `f(u)` 为激活函数的输出结果。
    -   `u` 为上一阶段计算出的加权和。

    换言之，若加权和大于等于 0，则感知机输出为 1；否则输出为 0。这种二分类输出
    可以表示是否推荐某款商品、是否将用户标记为潜在买家，或任何需要二值判断的目标事件。

3.  **Delta 规则（权重更新）：** 在使用激活函数计算出输出后，感知机会将该预测输出与
    目标输出（正确参考值）进行对比。两者的差值即为“误差”。为了最小化误差并实现“学习”，
    感知机采用 Delta 规则（Widrow-Hoff 学习规则）更新权重。该规则根据误差与对应输入按比例修正各权重。

    公式：
    `新权重 = 当前权重 + (学习率 * 误差 * 输入)`

    其中：

    -   `学习率`（learning rate）是一个控制学习步长速度的正数（详见“模型参数与训练数据”部分）。

为了在实践中直观展示感知机的运作机制，以下 Python 代码通过商业应用场景和类比实现了这一简单模型。

## 模型参数与训练数据

正如新员工需要明确的初始指引一样，感知机也需要初始参数与样本数据来启动学习过程。

In [1]:
# 感知机（Perceptron）神经网络实现
# Copyright (C) 2026 Renato Barros
#
# 本程序是自由软件：您可以根据自由软件基金会发布的 GNU 通用公共许可证（GNU General Public License）
# 条款重新分发和/或修改它，无论是许可证的第 3 版，还是（由您选择）任何更高版本。
#
# 本程序的发布是希望它能发挥作用，但没有任何保证；甚至没有对适销性或特定用途适用性的
# 暗示保证。详情请参阅 GNU 通用公共许可证。
#
# 您应该已经随本程序收到了一份 GNU 通用公共许可证的副本。如果没有，请参阅
# <https://www.gnu.org/licenses/>。

# 初始化感知机权重（w1 和 w2）
peso = [0.2, -0.1]
# 设定算法的学习率
taxa_aprendizagem = 0.1
# 初始化总误差计数器为 0
erros_totais = 0.0
# 设定最大训练轮数（Epochs）
limite_epocas = 20

# 定义训练数据集：包含输入特征（x1, x2）与期望目标标签（saida_desejada）
dados_treinamento = [
    {"x1": 0.5, "x2": 0.8, "saida_desejada": 1},
    {"x1": 0.2, "x2": 0.4, "saida_desejada": 0},
]

-   **初始化权重：** 在感知机中，“权重”代表各输入信息在决策中的重要程度。代码中
    `peso = [0.2, -0.1]` 表示为两个相关特征（x1 和 x2）分配了随机初始权重。

    例如，假设我们正在构建一个预测商品是否热销的模型。此时 x1 可以代表商品的好评数，
    x2 代表被加入心愿单的次数。权重（0.2 与 -0.1）反映了这些特征对最终判定的相对影响。

-   **设定学习率：** `taxa_aprendizagem = 0.1` 控制每次出现误差时感知机调整权重的幅度。
    可以将学习率类比为天平的灵敏度。

    较高的学习率会在每次出错时大幅调整权重，可能加速收敛但易导致震荡；
    较低的学习率则微调权重，训练过程平稳但耗时较长。

-   **累计误差：** `erros_totais = 0.0` 用作全局误差计数器。在训练期间，感知机对比
    预测值与真实标签，记录累计误差，帮助评估模型的学习效果。

    例如在电商潜在买家识别任务中，每当模型误判一次用户类别，计数器便会累加，
    直观反映出模型的准确水平。

-   **限制训练轮数：** `limite_epocas = 20` 设定了感知机遍历完整训练数据集的最大次数，
    用于控制训练耗时并避免死循环。

    在物流时效预测模型中，每一轮训练就相当于对所有历史物流记录进行一次完整复盘，
    使模型逐步优化权重以提升预测精度。

-   **提供训练数据：** `dados_treinamento` 是感知机学习的基础，包含了样本特征（x1, x2）
    及对应的正确分类标签（`saida_desejada = 1` 或 `0`）。

    如果用于热销品预测，训练集中会包含历史上热销的爆款案例与未成爆款的对照样本。

## 感知机训练

在设定好参数并准备好训练数据后，感知机开始进入学习流程。该过程通过迭代微调权重，
寻找最优参数组合以不断提升预测精度。

In [2]:
# 启动感知机训练循环
for epoca in range(1, limite_epocas + 1):
    print(f"\n Época: {epoca}\n")
    # 标记当前轮是否存在误差（初始设为 False）
    erro_na_epoca = False

    # 遍历每个训练样本
    for dado in dados_treinamento:
        # 提取输入特征与目标参考标签
        x1 = dado["x1"]
        x2 = dado["x2"]
        saida_desejada = dado["saida_desejada"]

        # 获取当前权重值
        w1 = peso[0]
        w2 = peso[1]

        # 计算感知机线性输出 (u)
        u = (x1 * peso[0]) + (x2 * peso[1])
        # 应用阶跃激活函数：
        # 若 u >= 0 则输出 1，否则输出 0
        valor_saida = 1 if u >= 0 else 0

        # 计算误差：目标期望值与预测输出的差值
        erro = saida_desejada - valor_saida

        # 累加绝对误差
        erros_totais += abs(erro)

        # 若当前样本分类存在误差，将标志位置为 True
        if erro != 0:
            erro_na_epoca = True
            # 使用 Delta 规则更新权重 (w1, w2)
            peso[0] += taxa_aprendizagem * erro * x1
            peso[1] += taxa_aprendizagem * erro * x2

        # 输出当前迭代详情（轮次、输入、预测、误差、权重状态）
        print(
            f" X1: {x1:.1f}, X2: {x2:.1f} | W1: {w1:.2f}, W2: {w2:.2f} | Y: {
                saida_desejada}, "
            f"Y-predito: {valor_saida} | Erro: {
                erro}, Erro acumulado: {erros_totais}"
        )
        if erro_na_epoca:
            print(
                f" Ajuste de pesos: Novo W1 = {
                    peso[0]:.2f}, Novo W2 = {peso[1]:.2f}\n"
            )
        if not erro_na_epoca:
            print(" Ajuste de pesos: Não foi necessário.\n")
    print("-=-" * 31)

    # 终止判定：若该轮整轮无误差（erro_na_epoca = False）
    if not erro_na_epoca:
        # 打印收敛成功信息与达成零误差的轮次
        print(
            f"\n Modelo treinado com sucesso na época {
                epoca}! Erro zero atingido."
        )
        # 打印最终最优权重值 (w1, w2)
        print(
            f" Valores ideais de peso: W1 = {
                peso[0]:.2f}, W2 = {peso[1]:.2f}\n"
        )
        # 结束训练循环
        break
# 若达到最大轮数仍未达到零误差
else:
    # 打印达到最大轮数提示
    print(
        f"\n O limite de {
            limite_epocas} épocas foi atingido sem alcançar erro zero."
    )
    # 提示感知机可能未能收敛
    print(" O perceptron não pôde ser treinado com sucesso.\n")


 Época: 1

 X1: 0.5, X2: 0.8 | W1: 0.20, W2: -0.10 | Y: 1, Y-predito: 1 | Erro: 0, Erro acumulado: 0.0
 Ajuste de pesos: Não foi necessário.

 X1: 0.2, X2: 0.4 | W1: 0.20, W2: -0.10 | Y: 0, Y-predito: 1 | Erro: -1, Erro acumulado: 1.0
 Ajuste de pesos: Novo W1 = 0.18, Novo W2 = -0.14

-=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=-

 Época: 2

 X1: 0.5, X2: 0.8 | W1: 0.18, W2: -0.14 | Y: 1, Y-predito: 0 | Erro: 1, Erro acumulado: 2.0
 Ajuste de pesos: Novo W1 = 0.23, Novo W2 = -0.06

 X1: 0.2, X2: 0.4 | W1: 0.23, W2: -0.06 | Y: 0, Y-predito: 1 | Erro: -1, Erro acumulado: 3.0
 Ajuste de pesos: Novo W1 = 0.21, Novo W2 = -0.10

-=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=-

 Época: 3

 X1: 0.5, X2: 0.8 | W1: 0.21, W2: -0.10 | Y: 1, Y-predito: 1 | Erro: 0, Erro acumulado: 3.0
 Ajuste de pesos: Não foi necessário.

 X1: 0.2, X2: 0.4 | W1: 0.21, W2: -0.10 | Y: 0, Y-predito: 1 | Erro: -1, Erro acumu

### 代码解析：

上述代码是感知机训练的核心，通过每轮迭代逐步优化模型权重以降低预测误差。
各步骤详细解析如下：

1.  **控制轮数（Epochs）：** 外层 `for` 循环控制训练轮数。每一轮代表感知机遍历所有训练样本，
    修正权重并统计总误差。最大轮数（20）限制了最大循环次数。

2.  **遍历训练样本：** 内层 `for` 循环逐一处理每个训练样本（x1, x2）及其正确标签（`saida_desejada`）。

3.  **提取样本信息：** 提取当前样本的输入特征与目标标签，用于计算输出及后续权重调整。

4.  **读取当前权重：** 变量 `w1` 与 `w2` 记录当前权重值，便于在训练过程中进行日志展示。

5.  **计算加权和：** 执行线性运算 `u = x1*w1 + x2*w2`。在爆款预测场景中，该加权和代表模型对商品的综合评分。

6.  **应用激活函数：** 阶跃函数根据加权和输出判定结果：若 `u >= 0` 则输出 1（正类），否则输出 0（负类）。

7.  **计算误差：** 误差等于目标真实值减去预测值 (`erro = saida_desejada - valor_saida`)。

8.  **调整权重（Delta 规则）：** 若预测存在误差，则依据 `taxa_aprendizagem * erro * x` 修正权重，
    使决策分界面逐步逼近正确分类。

9.  **输出轮次详情：** 实时打印当前轮次、输入特征、权重状态、预测值、误差及调整记录，全程透明可溯。

10. **收敛判定：** 每轮结束后检查该轮是否存在误差。若整轮无任何错误（`not erro_na_epoca`），
    则模型达成收敛，提前终止训练并展示最终最优权重。

感知机训练过程旨在寻找使误差归零的平衡点。这一简明而经典的机制，奠定了现代深度神经网络的基石。

## 结果分析与解释

分析模型运行输出日志，有助于我们评估感知机的学习过程与收敛质量。当模型在所有训练样本上均实现零误差时，
即标志着训练取得成功。

观察输出日志可见，感知机在第 12 轮达到了零误差收敛，表明模型经过 12 次样本遍历成功找到了能够完美划分数据的分类超平面。

经历多轮才达到零误差说明，尽管训练数据本身线性可分，但初始随机权重（`[0.2, -0.1]`）并非最优解，
必须通过多轮迭代反馈逐步微调修正。

最终训练得到的权重为 `W1 = 0.23`，`W2 = -0.14`，体现了模型从数据中学到的知识：

-   **W1 为正数 (+0.23)：** 表明特征 x1 对正类判定起正向推动作用。x1 数值越大，输出 1 的倾向越强。
-   **W2 为负数 (-0.14)：** 表明特征 x2 对正类判定起反向抑制作用。x2 数值越大，输出 1 的倾向越弱，与 x1 形成制衡。

此项分析验证了感知机综合权衡不同特征以学习复杂关系的能力。

需要注意的是，若最大轮数设为 3，训练将在未达成收敛前中断，此时得到的非最优权重将显著削弱模型的泛化能力。

## 总结

即使是最基础的单层感知机，也清晰展示了机器学习从数据中提取特征模式并解决分类问题的核心能力。
从客户细分到需求预测和安全筛查，其基本原理为构建更加智能化、敏捷化的现代决策系统提供了重要的理论与技术基础。